### DISCUSSION TITLE
First time jDeploy user: macOS installer flashes and exits??

### DISCUSSION THREAD

**Discussion by twolf2919** [shannah/jdeploy#111] [Q&A]
I have an application consisting of a main jar in dist/app.jar and some jars it depends on in the dist/lib/ directory.  I was using jpackage to create native installers and was hoping to move to jDeploy to solve my update problem.

I followed the developer guide and created an npm account and used the jDeploy GUI - started from my project's directory - to publish my application.  The app deploys successfully.  I then go to the download page and click on the Apple Silicon .app download button.  This causes a tar file to be downloaded and then extracted.  When I  open the installer in the extracted directory, I briefly see a splash screen with my application's title - and then the installer seems to exit.  I don't see any error messages in my Terminal when doing so.  I have no idea what's going on.  Is there a location where the installer writes any logging messages that can help me figure out what's going on?

Independently, does jDeploy have any provisions for specifying a EULA like jpackage?  I.e. My app is a commercial one, so I want the user to agree to terms of use before installing the app.  The jDeploy GUI has a  free-form"License" text field that defaults to "ISC".  Not knowing what the options are, I changed that to "Commercial" - but that's not really sufficient.

Many thanks for all information.

Here's the link to the download page in case that helps
https://www.jdeploy.com/~satriage

Follow-up: I just tried a couple of the demos (e.g. SwingSet2) and those do the same thing!?!  Could it be the fact that I'm running macOS 14 (Sonoma) beta?

**Comment by shannah** [comment]
Here's what I did to investigate.  Installed your app, then tried to run it from the command line:

```
cd ~/Applications/SATriage.app 
shannah@Steves-M1-mini SATriage.app % ./Contents/MacOS/Client4JLauncher 
[2023-08-21 14:21:40] [INFO   ] Initializing application 
Exception in thread "main" java.lang.NoClassDefFoundError: jakarta/xml/bind/JAXBContext
	at com.spectare.Context.recoverState(Context.java:408)
	at com.spectare.Startup.main(Startup.java:108)
	at com.spectare.gui.satriage.Main.main(Main.java:25)
Caused by: java.lang.ClassNotFoundException: jakarta.xml.bind.JAXBContext
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	... 3 more
2023/08/21 14:21:40 Failed to execute Java for package satriage version 1.0.1%!(EXTRA *errors.errorString=Java virtual machine exited with code 1)
Finished call to native%                   
```

This indicates that your app probably depends on a different version of Java than you have specified.
I see that you have `javaVersion` set to `17`, which doesn't include JAXB.  (They removed JAXB from the JDK in JDK 11 I think.).

So you either need to set javaVersion to 8, or you need to include the JAXB dependencies in your app.

In maven, this would mean adding something like this to your pom.xml file:

```
<!-- JAXB API -->
<dependency>
    <groupId>jakarta.xml.bind</groupId>
    <artifactId>jakarta.xml.bind-api</artifactId>
    <version>2.3.3</version>
</dependency>

<!-- JAXB runtime -->
<dependency>
    <groupId>org.glassfish.jaxb</groupId>
    <artifactId>jaxb-runtime</artifactId>
    <version>2.3.3</version>
</dependency>

```

**Reply by twolf2919** [reply]
Hi, thanks for giving the installer a try - but I don't even get to the post-install step as the installer exits immediately after showing the splash screen.  This also happens on every one of the samples your web site has listed.  Sonoma issue?

As far as the error you're seeing when trying to run my application:  I include those libraries in the "dist/" directory of my application.  I.e. my project/repo directory is called "app/" and I ran jDeploy from that directory and in the GUI pointed it to the main jar in dist/app.jar.  Since it's an executable jar, its Manifest points to quite a few jars in dist/lib/ directory - I thought jDeploy is supposed to find those, no?  I did see references to this discovery when I was trying to publish:
`Launching jdeploy gui.  Use jdeploy help for help
Includes: [CopyRule{dir=dist, includes=[app.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/commons-cli-1.4.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/commons-compress-1.20.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/gson-2.8.6.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jakarta.activation-2.0.1.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jakarta.xml.bind-api-4.0.0-RC2.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/javax.activation-1.2.0.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/javax.mail-1.6.2.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jaxb-core-3.0.2.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jaxb-impl-3.0.2.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jfreechart-1.5.3.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jgit-6.6.0.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jgit_jsch-6.6.0.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jna-5.10.0.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/jsch-0.1.55.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/lightning-csv-8.2.3.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/rsyntaxtextarea-3.1.6.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/sfm-csv-8.2.3.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/sfm-map-8.2.3.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/sfm-util-8.2.3.jar], excludes=null}, CopyRule{dir=dist, includes=[../lib/slf4j-api-2.0.7.jar], excludes=null}]
`
Maybe it didn't find them?  The "includes" weirdly refers to "../lib/" instead of "lib/" files when the dir is set to 'dist'.?

**Reply by shannah** [reply]
The install probably worked.
Look in `~/Applications/`.

(Note: That is not `/Applications` it is `~/Applications` - i.e. in your home directory).

**Reply by twolf2919** [reply]
How do I tell what jars were included in the publish process?  As per my other comment, there are 20 jars that should have been part of the application in addition to the main app.jar file.  I don't list them manually as dependencies in the package.json as the documentation indicated that they'd be discovered as part of  examining the executable jar's manifest file.  I definitely include the jakarta jaxb jar.

**Comment by shannah** [comment]
```
Follow-up: I just tried a couple of the demos (e.g. SwingSet2) and those do the same thing!?! Could it be the fact that I'm running macOS 14 (Sonoma) beta?
```

Interesting All of the demos work fine for me (on Ventura), but I don't have any Sonoma test boxes yet.  What happens when you try to run SwingSet via the command-line the way I described above?

**Reply by twolf2919** [reply]
When I run SwingSet2 via the command-line, I see the splash screen, it stays up a few seconds (but longer than when I invoke the app via Finder), then disappears and the command prompt returns.  Then a couple more seconds later, the actual installer screen shows up.

I then try the same thing  (invoke via command-line) with the SATriage installer and that works the same way?!?  Then I try invoking the installer directly from Finder (double-clicking on the .app) and now that doesn't even show the splash screen and immediately shows the installer screen.  So I'm baffled.

I went ahead and installed my app - and get the same error you pointed out above.  How do I tell if the jars that were in the dist/lib/ directory were actually included in the installer?  It's quite a few:
`twolf@bcd0-746b-8759 dist % ls lib
commons-cli-1.4.jar			jgit-6.6.0.jar
commons-compress-1.20.jar		jgit_jsch-6.6.0.jar
gson-2.8.6.jar				jna-5.10.0.jar
jakarta.activation-2.0.1.jar		jsch-0.1.55.jar
jakarta.xml.bind-api-4.0.0-RC2.jar	lightning-csv-8.2.3.jar
javax.activation-1.2.0.jar		rsyntaxtextarea-3.1.6.jar
javax.mail-1.6.2.jar			sfm-csv-8.2.3.jar
jaxb-core-3.0.2.jar			sfm-map-8.2.3.jar
jaxb-impl-3.0.2.jar			sfm-util-8.2.3.jar
jfreechart-1.5.3.jar			slf4j-api-2.0.7.jar`

Thanks again for your help.

p.s. is it possible to include an EULA (End user license agreement)  in the installer?  For a commercial application it's quite important to show this to the user before they use the product.  It's not a showstopper not to have that in the installer - I can put up a screen the first time the application is run - but the Installer would be the logical place.  It's an option in jpackage.

**Comment by shannah** [comment]
All of the packages are installed in `~/.jdeploy/packages`

I notice that your `Class-path` directive in your manifest references jars in the parent directory.  This might be a problem.

e.g.

```
cat META-INF/MANIFEST.MF
Manifest-Version: 1.0
Ant-Version: Apache Ant 1.10.14
Created-By: 17.0.5+9-LTS-191 (Oracle Corporation)
Main-Class: com.spectare.gui.satriage.Main
Class-Path: ../lib/commons-cli-1.4.jar ../lib/commons-compress-1.20.ja
 r ../lib/gson-2.8.6.jar ../lib/jakarta.activation-2.0.1.jar ../lib/ja
 karta.xml.bind-api-4.0.0-RC2.jar ../lib/javax.activation-1.2.0.jar ..
 /lib/javax.mail-1.6.2.jar ../lib/jaxb-core-3.0.2.jar ../lib/jaxb-impl
 -3.0.2.jar ../lib/jfreechart-1.5.3.jar ../lib/jgit-6.6.0.jar ../lib/j
 git_jsch-6.6.0.jar ../lib/jna-5.10.0.jar ../lib/jsch-0.1.55.jar ../li
 b/lightning-csv-8.2.3.jar ../lib/rsyntaxtextarea-3.1.6.jar ../lib/sfm
 -csv-8.2.3.jar ../lib/sfm-map-8.2.3.jar ../lib/sfm-util-8.2.3.jar ../
 lib/slf4j-api-2.0.7.jar
```

Try changing your build so that you bundle your dependent jars at the same level or below your app jar.

E.g.

```
app.jar
   +--- lib/commons-cli-1.4.jar
   .... etc...
 ```
 
 > p.s. is it possible to include an EULA (End user license agreement) in the installer? For a commercial application it's quite important to show this to the user before they use the product. It's not a showstopper not to have that in the installer - I can put up a screen the first time the application is run - but the Installer would be the logical place. It's an option in jpackage.
 
 Not currently, unless you embedded it in your installer image.
 
 You could include such an EULA the first time your app is launched (e.g. embed that stuff in the app itself).

**Reply by twolf2919** [reply]
Thanks a bunch - the class path references using the '..' was the problem.  Now everything works as expected.
